### PR TITLE
Add support for Neo N3

### PR DESCRIPTION
**Pull request by joeqian10**

### CODE REVIEW

**Comment by joeqian10** [comment]
@zhiqiangxu @tanZiWen Please review the code before merging. Thanks!

File: native/service/governance/neo3_state_manager/utils.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+import (
+	"fmt"
+	"github.com/polynetwork/poly/common"
+	cstates "github.com/polynetwork/poly/core/states"
+	"github.com/polynetwork/poly/native"
+	"github.com/polynetwork/poly/native/event"
+	"github.com/polynetwork/poly/native/service/utils"
+)
+
+//// putStateValidator put state validator in the contract storage using the STATE_VALIDATOR key
```
Comment (tanZiWen) [inline native/service/governance/neo3_state_manager/utils.go:12]: remove unused code

File: native/service/governance/neo3_state_manager/utils.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+	return sink.Bytes()
+}
+
+func DeserializeStringArray(data []byte) ([]string, error) {
+	source := common.NewZeroCopySource(data)
+	n, eof := source.NextVarUint()
+	if eof {
+		return nil, fmt.Errorf("source.NextVarUint error")
+	}
+	result := make([]string, 0)
```
[the remark this answers is not in the document]
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/utils.go]: updated

File: native/service/governance/neo3_state_manager/utils.go (hunk 1 above)
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/utils.go]: removed

File: native/service/cross_chain_manager/neo3/neo_handler.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	if err := scom.CheckDoneTx(service, value.CrossChainID, params.SourceChainID); err != nil {
+		return nil, fmt.Errorf("neo3 MakeDepositProposal, check done transaction error:%s", err)
+	}
+	if err = scom.PutDoneTx(service, value.CrossChainID, params.SourceChainID); err != nil {
+		return nil, fmt.Errorf("neo3 MakeDepositProposal, putDoneTx error:%s", err)
+	}
+	return value, nil
+}
+
+func (this *Neo3Handler) ConvertCcmcAddressToId(idBytes []byte) int {
```
Comment (tanZiWen) [inline native/service/cross_chain_manager/neo3/neo_handler.go:77]: remove

File: native/service/cross_chain_manager/neo3/neo_handler.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	if err := neo3.VerifyCrossChainMsgSig(service, helper.BytesToUInt32(sideChain.ExtraInfo), crossChainMsg); err != nil {
+		return nil, fmt.Errorf("neo3 MakeDepositProposal, VerifyCrossChainMsg error: %v", err)
+	}
+
+	// when register neo N3, convert ccmc id to []byte
+	// convert neo3 contract address bytes to id, it is different from other chains
+	// need to store int in a []byte, contract id can be get from "getcontractstate" api
+	// neo3 native contracts have negative ids, while custom contracts have positive ones
+
+	id :=	this.ConvertCcmcAddressToId(sideChain.CCMCAddress)
```
Comment (tanZiWen) [inline native/service/cross_chain_manager/neo3/neo_handler.go:61]: id: =int(int32(helper.BytesToUInt32(idBytes)))

File: native/service/governance/neo3_state_manager/neo3_state_manager.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	ok, err := node_manager.CheckConsensusSigns(native, APPROVE_REGISTER_STATE_VALIDATOR, utils.GetUint64Bytes(params.ID), params.Address)
+	if err != nil {
+		return utils.BYTE_FALSE, fmt.Errorf("ApproveRegisterStateValidator, CheckConsensusSigns error: %v", err)
+	}
+	if !ok {
+		return utils.BYTE_FALSE, nil
+	}
+
+	// put all the state validators in storage
+	err = putStateValidators(native, svListParam.StateValidators)
```
Comment (tanZiWen) [inline native/service/governance/neo3_state_manager/neo3_state_manager.go:90]: whether need to remove applyID?

File: native/service/governance/neo3_state_manager/utils.go [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+func putStateValidators(native *native.NativeService, stateValidators []string) error {
+	contract := utils.Neo3StateManagerContractAddress
+	// get current stored value
+	oldSVs, err := getStateValidators(native)
+	if err != nil {
+		return fmt.Errorf("putStateValidator, get old state validators error: %v", err)
+	}
+	// use a map to filter old svs
+	mm := make(map[string]string)
```
Comment (tanZiWen) [inline native/service/governance/neo3_state_manager/utils.go:67]: why use map remove the duplicate key, the golang map is out-of-order, so if you use map, every node will run different result, you also store the result in the db, the state root maybe not same.

File: native/service/header_sync/neo3/utils.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	// todo, review code, get neo3 state validator from native contract
+	svListBytes, err := neo3_state_manager.GetCurrentStateValidator(native)
+	if err != nil {
+		return fmt.Errorf("verifyCrossChainMsg, neo3_state_manager.GetCurrentStateValidator error: %v", err)
+	}
+	svStrings, err := neo3_state_manager.DeserializeStringArray(svListBytes)
+	if err != nil {
+		return fmt.Errorf("verifyCrossChainMsg, neo3_state_manager.DeserializeStringArray error: %v", err)
+	}
+	pubKeys := make([]crypto.ECPoint, len(svStrings))
```
Comment (tanZiWen) [inline native/service/header_sync/neo3/utils.go:64]: Why you store the public key on db? it need convert string to public key every times.

File: native/service/governance/neo3_state_manager/neo3_state_manager.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+)
+
+// todo, review code
+
+func GetCurrentStateValidator(native *native.NativeService) ([]byte, error) {
+	svs, err := getStateValidators(native)
+	if err != nil {
+		return utils.BYTE_FALSE, fmt.Errorf("GetCurrentStateValidator, getStateValidators error: %v", err)
+	}
+	data := SerializeStringArray(svs)
```
Comment (tanZiWen) [inline native/service/governance/neo3_state_manager/neo3_state_manager.go:35]: why this place need to serialize it？

File: native/service/header_sync/neo3/utils.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	return nil
+}
+
+func VerifyCrossChainMsgSig(native *native.NativeService, magic uint32, crossChainMsg *NeoCrossChainMsg) error {
+	// todo, review code, get neo3 state validator from native contract
+	svListBytes, err := neo3_state_manager.GetCurrentStateValidator(native)
+	if err != nil {
+		return fmt.Errorf("verifyCrossChainMsg, neo3_state_manager.GetCurrentStateValidator error: %v", err)
+	}
+	svStrings, err := neo3_state_manager.DeserializeStringArray(svListBytes)
```
Comment (tanZiWen) [inline native/service/header_sync/neo3/utils.go:60]: you serialize it in `GetCurrentStateValidator` function and deserialize it again, why?

File: native/service/governance/neo3_state_manager/param.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
+package neo3_state_manager
+
+import (
+	"fmt"
+	"github.com/polynetwork/poly/common"
+)
+
+type StateValidatorListParam struct {
+	StateValidators []string       // public key strings in encoded format, each is 33 bytes in []byte
+	Address         common.Address // for check witness?
```
Comment (tanZiWen) [inline native/service/governance/neo3_state_manager/param.go:10]: which place will use `Address`?

File: native/service/governance/neo3_state_manager/param.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+		return fmt.Errorf("common.AddressParseFromBytes, deserialize address error: %s", err)
+	}
+	this.StateValidators = stateValidators
+	this.Address = addr
+	return nil
+}
+
+type ApproveStateValidatorParam struct {
+	ID      uint64         // StateValidatorApproveID
+	Address common.Address // for check witness?
```
Comment (tanZiWen) [inline native/service/governance/neo3_state_manager/param.go:50]: which place will use `Address`?

File: native/service/governance/neo3_state_manager/param.go (hunk 10 above)
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/param.go:10]: I am not sure, maybe when calling the contract methods. `RelayerListParam` uses `Address`:
```go
type RelayerListParam struct {
	AddressList []common.Address
	Address     common.Address
}
```

File: native/service/governance/neo3_state_manager/param.go (hunk 11 above)
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/param.go:50]:
```go
type ApproveRelayerParam struct {
	ID      uint64
	Address common.Address
}
```

File: native/service/governance/neo3_state_manager/neo3_state_manager.go (hunk 8 above)
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/neo3_state_manager.go:35]: Since a `Handler` can only return []byte.

Comment (joeqian10) [reply native/service/governance/neo3_state_manager/neo3_state_manager.go:35]:
```go
Handler func(native *NativeService) ([]byte, error)
```

File: native/service/cross_chain_manager/neo3/neo_handler.go (hunk 3 above)
Comment (joeqian10) [reply native/service/cross_chain_manager/neo3/neo_handler.go]: removed

File: native/service/governance/neo3_state_manager/utils.go [hunk 12]
```diff
@@ -0,0 +1,247 @@
+package neo3_state_manager
```
Comment (tanZiWen) [inline native/service/governance/neo3_state_manager/utils.go:1]: add license

File: native/service/governance/neo3_state_manager/neo3_state_manager.go [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+	if !ok {
+		return utils.BYTE_FALSE, nil
+	}
+
+	// remove svs
+	err = removeStateValidators(native, svListParam.StateValidators)
+	if err != nil {
+		return utils.BYTE_FALSE, fmt.Errorf("ApproveRemoveStateValidator, removeStateValidators error: %v", err)
+	}
+
```
Comment (tanZiWen) [inline native/service/governance/neo3_state_manager/neo3_state_manager.go:151]: After finished remove validators, why not delete the removeid from db?

File: native/service/cross_chain_manager/neo3/neo_handler.go (hunk 4 above)
Comment (joeqian10) [reply native/service/cross_chain_manager/neo3/neo_handler.go]: udpated

File: native/service/governance/neo3_state_manager/utils.go (hunk 6 above)
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/utils.go]: udpated

File: native/service/governance/neo3_state_manager/neo3_state_manager.go (hunk 8 above)
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/neo3_state_manager.go]: udpated

File: native/service/governance/neo3_state_manager/utils.go (hunk 12 above)
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/utils.go:18]: udpated

File: native/service/governance/neo3_state_manager/neo3_state_manager.go (hunk 13 above)
Comment (joeqian10) [reply native/service/governance/neo3_state_manager/neo3_state_manager.go:165]: udpated

**Pull request merged**